src/io_darwin.zig

```diff
@@ -0,0 +1,691 @@
+            }
+
+            const new_events = try os.kevent(
+                self.kq, 
+                events[0..change_events],
+                events[0..events.len],
+                &ts,
+            );
+
+            // Mark the io events submitted only after kevent() successfully processed them
+            self.io_pending.out = io_pending;
+            if (io_pending == null) {
+                self.io_pending.in = null;
+            }
+
+            for (events[0..new_events]) |event| {
+                const completion = @intToPtr(*Completion, event.udata);
+                self.completed.push(completion);
+            }
+        }
+
+        var completed = self.completed;
+        self.completed = .{};
+        while (completed.pop()) |completion| {
+            (completion.callback)(self, completion);
+        }
+    }
+
+    fn flush_io(self: *IO, events: []os.Kevent, io_pending_top: *?*Completion) usize {
+        for (events) |*event, flushed| {
+            const completion = io_pending_top.* orelse return flushed;
+            io_pending_top.* = completion.next;
+
+            const event_info = switch (completion.operation) {
+                .accept => |op| [2]c_int{ op.socket, os.EVFILT_READ },
+                .connect => |op| [2]c_int{ op.socket, os.EVFILT_WRITE },
+                .read => |op| [2]c_int{ op.fd, os.EVFILT_READ },
+                .write => |op| [2]c_int{ op.fd, os.EVFILT_WRITE },
+                .recv => |op| [2]c_int{ op.socket, os.EVFILT_READ },
+                .send => |op| [2]c_int{ op.socket, os.EVFILT_WRITE },
+                else => @panic("invalid completion operation queued for io"),
+            };
+
+            event.* = .{
+                .ident = @intCast(u32, event_info[0]),
+                .filter = @intCast(i16, event_info[1]),
+                .flags = os.EV_ADD | os.EV_ENABLE | os.EV_ONESHOT,
+                .fflags = 0,
+                .data = 0,
+                .udata = @ptrToInt(completion),
+            };
+        }
+        return events.len;
+    }
+    
+    fn flush_timeouts(self: *IO) ?u64 {
+        var min_timeout: ?u64 = null;
+        var timeouts: ?*Completion = self.timeouts.peek();
+        while (timeouts) |completion| {
+            timeouts = completion.next;
+
+            // NOTE: We could cache `now` above but monotonic() should be cheap to call ideally.
+            const now = self.time.monotonic();
+            const expires = completion.operation.timeout.expires;
+            
+            // NOTE: remove could be O(1) here since we could know the previous Completion
+            if (now >= expires) {
+                self.timeouts.remove(completion);
+                self.completed.push(completion);
+                continue;
+            }
+
+            const timeout_ns = expires - now;
+            const current_ns = min_timeout orelse std.math.maxInt(u64);
+            min_timeout = std.math.min(current_ns, timeout_ns);
```

Comment:
Let's rather use the optional of `min_timeout` to decide whether to take the min, rather than always taking the min and introducing `maxInt(u64)` because that opens a TOCTOU or semantic gap if we ever change the type. I'm thinking `null` is always a better indicator of "not-set" than using a max int. It's a little more verbose but that's fine.